### PR TITLE
Add query to ensure predicates starting with 'get' return a value

### PR DESCRIPTION
**Pull request by Napalys**
This PR adds a new query to ensure that predicates starting with 'get' actually return a value. It includes the query definition, expected test results, and test cases to validate the functionality.

### CODE REVIEW

File: ql/ql/src/queries/style/ValidatePredicateGetReturns.ql [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+/**
+ * Checks if a predicate has a return type.
+ */
+predicate hasReturnType(Predicate pred) { exists(pred.getReturnType()) }
+
+/**
+ * Checks if a predicate is an alias using getAlias().
+ */
+predicate isAlias(Predicate pred) {
+  pred instanceof ClasslessPredicate and exists(pred.(ClasslessPredicate).getAlias())
```
Comment (aschackmull) [inline ql/ql/src/queries/style/ValidatePredicateGetReturns.ql:30]: The instanceof check is redundant.
```suggestion
  exists(pred.(ClasslessPredicate).getAlias())
```

File: ql/ql/src/queries/style/ValidatePredicateGetReturns.ql [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+import ql
+import codeql_ql.ast.Ast
+
+/**
+ * Identifies predicates whose names start with "get" followed by an uppercase letter.
+ * This ensures that only predicates like "getValue" are matched, excluding names like "getter".
+ */
+predicate isGetPredicate(Predicate pred) { pred.getName().regexpMatch("get[A-Z].*") }
```
Comment (aschackmull) [inline ql/ql/src/queries/style/ValidatePredicateGetReturns.ql:19]: We could also include the `as` prefix:
```suggestion
predicate isGetPredicate(Predicate pred) { pred.getName().regexpMatch("(get|as)[A-Z].*") }
```

File: ql/ql/test/queries/style/ValidatePredicateGetReturns/ValidatePredicateGetReturns.expected [hunk 3]
```diff
@@ -0,0 +1,3 @@
+| test.qll:4:11:4:18 | ClasslessPredicate getValue | This predicate starts with 'get' but does not return a value. |
+| test.qll:25:11:25:28 | ClasslessPredicate getImplementation2 | This predicate starts with 'get' but does not return a value. |
+| test.qll:31:11:31:17 | ClasslessPredicate asValue | This predicate starts with 'get' but does not return a value. |
```
Comment (aschackmull) [inline ql/ql/test/queries/style/ValidatePredicateGetReturns/ValidatePredicateGetReturns.expected:3]:
```suggestion
| test.qll:31:11:31:17 | ClasslessPredicate asValue | This predicate starts with 'as' but does not return a value. |
```

File: ql/ql/src/queries/style/ValidatePredicateGetReturns.ql [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+ * Checks if a predicate is an alias using getAlias().
+ */
+predicate isAlias(Predicate pred) { exists(pred.(ClasslessPredicate).getAlias()) }
+
+from Predicate pred
+where
+  isGetPredicate(pred) and
+  not hasReturnType(pred) and
+  not isAlias(pred)
+select pred, "This predicate starts with 'get' but does not return a value."
```
[the remark this answers is not in the document]
Comment (Napalys) [reply ql/ql/src/queries/style/ValidatePredicateGetReturns.ql]: a462ec91f5d6e8e881f370a6823202f8c6f1c400 🙈

File: ql/ql/src/queries/style/ValidatePredicateGetReturns.ql [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+/**
+ * Checks if a predicate is an alias using getAlias().
+ */
+predicate isAlias(Predicate pred) { exists(pred.(ClasslessPredicate).getAlias()) }
+
+/**
+ * Returns "get" if the predicate name starts with "get", otherwise "as".
+ */
+string getPrefix(Predicate pred) {
```
Comment (aschackmull) [inline ql/ql/src/queries/style/ValidatePredicateGetReturns.ql:34]: There's really no need for this predicate - just replace `regexpMatch` with `regexpCapture` in `isGetPredicate` and add the prefix column to that predicate.

Comment (Napalys) [reply ql/ql/src/queries/style/ValidatePredicateGetReturns.ql]: I see, that does make it cleaner!
67745e6332564b343537ab63e08330a056eeed4f

File: ql/ql/src/queries/style/ValidatePredicateGetReturns.ql [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+/**
+ * Checks if a predicate is an alias using getAlias().
+ */
+predicate isAlias(Predicate pred) { exists(pred.(ClasslessPredicate).getAlias()) }
+
+from Predicate pred, string prefix
+where
+  isGetPredicate(pred, prefix) and
+  not hasReturnType(pred) and
+  not isAlias(pred)
```
Comment (hvitved) [inline ql/ql/src/queries/style/ValidatePredicateGetReturns.ql:37]: Should we instead check that whatever it aliases has a return type?

**Review by aschackmull — APPROVED** [review]

File: ql/ql/src/queries/style/ValidatePredicateGetReturns.ql (hunk 6 above)
Comment (aschackmull) [reply ql/ql/src/queries/style/ValidatePredicateGetReturns.ql:37]: I think that's a very reasonable extension of this query, but I don't mind merging as-is, hence approved.

File: ql/ql/test/queries/style/ValidatePredicateGetReturns/test.qll [hunk 7]
```diff
@@ -40,3 +40,28 @@ string asString() { result = "string" }
 HiddenType getInjectableCompositeActionNode() {
   exists(HiddenType hidden | result = hidden.toString())
 }
+
+// OK -- starts with get and does not return value
```
Comment (hvitved) [inline ql/ql/test/queries/style/ValidatePredicateGetReturns/test.qll:44]: doesn't start with get

File: ql/ql/test/queries/style/ValidatePredicateGetReturns/test.qll [hunk 8]
```diff
@@ -40,3 +40,28 @@ string asString() { result = "string" }
...
+// OK -- Higher-order predicate
+int getDistance(int x) = shortestDistances(root/0, edge/2)(_, x, result)
+
+// NOT OK -- Higher-order predicate that does not return a value even though has 'get' in the name
+predicate getDistance2(int x, int y) = shortestDistances(root/0, edge/2)(_, x, y)
+
+// OK
+predicate unresolvedAlias = unresolved/0;
+
+// NOT OK -- unresolved alias
```
Comment (hvitved) [inline ql/ql/test/queries/style/ValidatePredicateGetReturns/test.qll:66]: This may or may not be OK, so the query shouldn't flag it (which is doesn't).

**Review by hvitved — CHANGES_REQUESTED** [review]

File: ql/ql/src/queries/style/ValidatePredicateGetReturns.ql (hunk 6 above)
Comment (hvitved) [reply ql/ql/src/queries/style/ValidatePredicateGetReturns.ql]: Yes, thanks for adding.

File: ql/ql/test/queries/style/ValidatePredicateGetReturns/test.qll (hunk 7 above)
Comment (Napalys) [reply ql/ql/test/queries/style/ValidatePredicateGetReturns/test.qll]: Sorry that was sloppy

**Review by hvitved — APPROVED** [review]

**Pull request merged**